### PR TITLE
Add support for tagging

### PR DESCRIPTION
**Pull request by Serphentas**
Basically making it possible to use [tags](https://graphite.readthedocs.io/en/latest/tags.html), nothing much to it

Would be nice to have a flattening function if tags are nested

### CODE REVIEW

File: graphyte.py [hunk 1]
```diff
@@ -64,31 +64,41 @@ def stop(self):
             self._thread.join()
             self.interval = None
 
-    def build_message(self, metric, value, timestamp):
+    def build_message(self, metric, value, timestamp, tags):
```
Comment (benhoyt) [inline graphyte.py]: This is a "public" function (no underscore prefix), so I'd prefer it if the signature was backwards-compatible (that's the reason it broke the tests). For example, if folks are subclassing and relying on the existing signature, it'll break for them. Can we make this have an empty default as well, like `tags=[]`?

File: graphyte.py [hunk 2]
```diff
@@ -64,31 +64,41 @@ def stop(self):
...
         if not isinstance(value, (int, float)):
             raise TypeError('"value" must be an int or a float, not a {}'.format(
-                    type(value).__name__))
-
-        message = u'{}{} {} {}\n'.format(
+                type(value).__name__))
+        if not (isinstance(tags, list) and all(isinstance(x, dict) for x in tags)):
+            raise TypeError('"tags" must be a list of dicts')
+        if not (all(all(isinstance(key, str) for key in tag.keys()) for tag in tags)):
+            raise TypeError('tag key must be an str')
```
Comment (benhoyt) [inline graphyte.py]: A couple of things here:

1) Why isn't the tags parameter simply a dict, rather than a list of dicts? Then you could just pass it as `tags={'tag1': 'value1', 'tag2': 'value2'}`. I don't see why the API should support multiple dicts of tags, and if the caller has that shape, they can merge before calling.
2) I realize we have a type check for the `value`, though that's a bit special. In general Python encourages "duck typing", meaning you don't check specific types but expect it to act like a duck (dict in this case). So I think we should just remove these type checks.

File: graphyte.py [hunk 3]
```diff
@@ -64,31 +64,41 @@ def stop(self):
...
+
+        tags_suffix = ''
+        for tag in tags:
+            for key in tag.keys():
+              tags_suffix += ';{}={}'.format(key, tag[key])
```
Comment (benhoyt) [inline graphyte.py]: It's good practice to use `str.join` rather than repeated appending -- strings are immutable in Python, so this usually recreates a new string object every time around the loop causing O(N^2) behaviour. In this case it probably doesn't matter (small number of tags), but I'd prefer good practice. So, assuming we change tags to a dict, this would become:

    tags_suffix = ''.join(';{}={}'.format(k, v) for k, v in tags.items())

File: graphyte.py [hunk 4]
```diff
@@ -64,31 +64,41 @@ def stop(self):
...
+
+        message = u'{}{} {} {}{}\n'.format(
             self.prefix + '.' if self.prefix else '',
             metric,
             value,
             int(round(timestamp)),
+            tags_suffix
```
Comment (benhoyt) [inline graphyte.py]: Are you sure Carbon wants the tags suffix at the end of the line, rather than as part of the metric name? From the docs page you linked, and from the go-carbon code I was looking at, it looked like the tags suffix should be at the end of the metric name ... but I could be wrong here.

File: graphyte.py [hunk 5]
```diff
@@ -64,31 +64,41 @@ def stop(self):
...
         )
         message = message.encode('utf-8')
         return message
 
-    def send(self, metric, value, timestamp=None):
+    def send(self, metric, value, timestamp=None, tags=[]):
```
Comment (benhoyt) [inline graphyte.py]: This will become `tags={}` if tags is a dict. We should also add a note about the new `tags` param and the format of it in the docstring, maybe:

    If a "tags" dict is specified, send the tags to the Graphite host with the metric.

File: test_graphyte.py [hunk 6]
```diff
@@ -43,27 +43,27 @@ def pop_message(cls):
 class TestBuildMessage(unittest.TestCase):
     def test_no_prefix(self):
         sender = TestSender()
-        self.assertEqual(sender.build_message('foo.bar', 42, 12345),
+        self.assertEqual(sender.build_message('foo.bar', 42, 12345, []),
```
Comment (benhoyt) [inline test_graphyte.py]: These tests should be reverted once the `tags` param on build_message is changed to have a default.

In addition, we should add some tags-specific tests: at least 1) empty dict of tags, 2) one tag, 3) multiple tags

**Review by benhoyt — COMMENTED** [review]
Thanks for the PR! I didn't know about this feature. I've left a couple of comments as well as some questions.

File: graphyte.py (hunk 4 above)
Comment (Serphentas) [reply graphyte.py]: Indeed, it says [here](https://graphite.readthedocs.io/en/latest/tags.html#tags) that it should be appended to the series name. Will fix that.

File: graphyte.py (hunk 3 above)
Comment (Serphentas) [reply graphyte.py]: Indeed it probably won't impact performance that much but implementing BCPs is always good, plus one-liners are somewhat easier to understand.

File: graphyte.py (hunk 2 above)
Comment (Serphentas) [reply graphyte.py]: 1. Agreed, it's actually much simpler in many aspects
2. That was just to ensure proper tag format before sending them to Carbon, but anyhow if it fails then the metric will just be dropped.

File: graphyte.py (hunk 5 above)
Comment (Serphentas) [reply graphyte.py]: Done

File: graphyte.py (hunk 1 above)
Comment (Serphentas) [reply graphyte.py]: Done, but with a default dict type.

**Comment by Serphentas** [comment]
I've resolved the issues you raised, but it seems that tagging tests fail on Python 2.7 for some reason.

**Comment by Serphentas** [comment]
Okay fixed, we're all set now.

File: graphyte.py [hunk 7]
```diff
@@ -64,31 +64,39 @@ def stop(self):
...
-    def build_message(self, metric, value, timestamp):
+    def build_message(self, metric, value, timestamp, tags={}):
         """Build a Graphite message to send and return it as a byte string."""
         if not metric or metric.split(None, 1)[0] != metric:
             raise ValueError('"metric" must not have whitespace in it')
         if not isinstance(value, (int, float)):
             raise TypeError('"value" must be an int or a float, not a {}'.format(
-                    type(value).__name__))
+                type(value).__name__))
+        if not isinstance(tags, dict):
```
Comment (benhoyt) [inline graphyte.py]: I'd prefer to remove this `isinstance()` check entirely -- in the spirit of duck typing, let's just allow any dict-like object (in this case anything that supports `.items()`).

File: graphyte.py [hunk 8]
```diff
@@ -64,31 +64,39 @@ def stop(self):
...
+            raise TypeError('"tags" must be a dict, not a {}'.format(
+                type(tags).__name__))
 
-        message = u'{}{} {} {}\n'.format(
+        tags_suffix = ''.join(';{}={}'.format(x[0], x[1]) for x in sorted(tags.items()))
```
Comment (benhoyt) [inline graphyte.py]: Looks good -- nice idea to sort the order of the tags to produce consistent output and make testing easier.

**Review by benhoyt — COMMENTED** [review]
Thanks -- just one more comment about removing the isinstance check and then looks good to me.

File: graphyte.py (hunk 7 above)
Comment (Serphentas) [reply graphyte.py]: Done. If anything other than a dict-like object is passed, it will simply fail at the suffix generation step.

File: graphyte.py (hunk 8 above)
Comment (Serphentas) [reply graphyte.py]: Only way to make things work across all release versions, I guess.

**Comment by Serphentas** [comment]
Just implemented the last needed changes.

**Comment by benhoyt** [comment]
Excellent, thanks -- will merge!

**Comment by benhoyt** [comment]
I just uploaded 1.5 with this change to PyPI as well: https://pypi.org/project/graphyte/

**Comment by Serphentas** [comment]
Cool, thanks a bunch ! This library just got ten times better.
